### DISCUSSION TITLE
Not Beginner Friendly Project

### DISCUSSION THREAD

**Discussion by ThePieMonster** [fastapi/full-stack-fastapi-template#1115] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched in the documentation/README.
- [X] I already searched in Google "How to do X" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs/README and didn't find an answer.

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

After spending a few hours on this project and it still not building I've noticed that there is very little documentation that explains how the Docker images are setup. 

For example I have a test failing cause files are missing, and staging deployment failing cause some networking issue. Where do the tests come from? What is already being included in the backend docker image? 

Its better to use a raw Ubuntu image and show the steps being taken to build the image that is being used to run these other tests.

### Description

See above

### Operating System

Windows

### Operating System Details

See above

### Python Version

See above

### Additional Context

See above

**Comment by Happily-Coding** [comment, marked as answer]
I agree that this is not a very beginner friendly project, but keep in mind the idea for this project is not to be beginner friendly, but to use best practices and provide a good base developers can start with.

That said, I'm a bit surprised that you spent hours trying to build the project, these steps should be enough:
1. Download and install docker desktop.
2. Run docker desktop.
3. Download the repository, and extract it to a folder.
4. Open a command line terminal on the repository base folder.
5. Run docker compose up --build
Your page should be available on http://localhost

## Concerning **docker** this is structure as I understand it:
The frontend and backend each have their own docker file, indicating how to create their own image.
Docker-compose is used to build both of them, along with some other existing images, and allow the images to correctly communicate with each other.

There are multiple docker-compose files:
- **docker-compose.yml** provides the **base** docker-compose configuration, commmon for both development and production. In deployment its used as is without being overriden, by calling `docker compose -f docker-compose.yml up -d`
- **docker-compose.override** extends and overrides **docker-compose.yml** . Its used for **development**, and will be used alongside docker-compose if you simply run `docker compose up --build`.
- **docker.compose.traefik** is enabled independantly of the main docker compose, and should be used alongside **docker-compose.yml** in **production deployment**.  docker-compose wont use it by default. Its my understanding that it might be possible to include it in a single command as an additional file: docker-`compose -f docker-compose.yml -f docker-compose.traefik.yml up --build`. But thats not how it was intended to be used. See [here for more details](https://github.com/tiangolo/full-stack-fastapi-template/blob/master/deployment.md) . I believe this may be because you may want to run a single traefik server connected to multiple applications, and therefore you want traefik to be enabled separately.



**_Please if you agree or disagree with my interpretation of the structure leave a comment, so we can know with more certainty_**

---
For something a bit more beginner friendly you might wanna check one of these out:
https://github.com/Happily-Coding/FastapiReactNginxDocker https://github.com/jonasrenault/fastapi-react-mongodb-docker

**Reply by cetcrack** [reply]
Hello @Happily-Coding, im very new and wanted help with using this template
how do i make changes to the code and have live reload?
do i have to build new image after every change ? :smile: :smile:

**Reply by Happily-Coding** [reply]
I'm still familiarizing myself with the template, but, as of my current understanding with the current template you cannot take advantage of hot reload on the dockerized frontend.
Instead, use cd frontend and then npm run dev to run the frontend with hot reload, please see https://github.com/tiangolo/full-stack-fastapi-template/tree/master/frontend

It wouldn't be impossible to change the template to allow hot reloading in the dockerized frontend, and I'd seriously prefer that except for deployment, frontend hotreloading worked in dockerized form, but that doesn't currently seem to be the case.

**Reply by yaoqiang** [reply]
> I'm still familiarizing myself with the template, but, as of my current understanding with the current template you cannot take advantage of hot reload on the dockerized frontend. Instead, use cd frontend and then npm run dev to run the frontend with hot reload, please see https://github.com/tiangolo/full-stack-fastapi-template/tree/master/frontend
> 
> It wouldn't be impossible to change the template to allow hot reloading in the dockerized frontend, and I'd seriously prefer that except for deployment, frontend hotreloading worked in dockerized form, but that doesn't currently seem to be the case.

Hi, @Happily-Coding , I have same problerm in frontend development and deployment, are you development local with npm run dev, I think it's ok, but how did you deploy with docker? I am try to exec `docker compose up -d` when frontend code updated, but the docker image not update. I have to exec `docker compose down` and up again.  so, did you have some sugguest for me? thx

**Reply by QDFish** [reply]
> > I'm still familiarizing myself with the template, but, as of my current understanding with the current template you cannot take advantage of hot reload on the dockerized frontend. Instead, use cd frontend and then npm run dev to run the frontend with hot reload, please see https://github.com/tiangolo/full-stack-fastapi-template/tree/master/frontend
> > It wouldn't be impossible to change the template to allow hot reloading in the dockerized frontend, and I'd seriously prefer that except for deployment, frontend hotreloading worked in dockerized form, but that doesn't currently seem to be the case.
> 
> Hi, @Happily-Coding , I have same problerm in frontend development and deployment, are you development local with npm run dev, I think it's ok, but how did you deploy with docker? I am try to exec `docker compose up -d` when frontend code updated, but the docker image not update. I have to exec `docker compose down` and up again. so, did you have some sugguest for me? thx

`docker compose up -d --build` would solve your problem, it will rebuild using dockerfile
